### ISSUE TITLE
integer expression expected error

### ISSUE THREAD

**Issue by yoasif** [dim-geo/partialscrub#1]
Saw this upon running this on my array:

```
root@butter-server:/home/asif/bin/partialscrub# ./partialscrub.sh -d /media/camino/
filesystem:/media/camino/
raid56:on
percentage:100
frequency:2629744
scrub must be restarted:0
WARNING: RAID56 detected, not implemented
total data:42308876570624
data scrubbed:4423790592
status:aborted
limit:42313300361216
active device:0
scrub resumed on /dev/sdh, fsid 48ed8a66-731d-499b-829e-dd07dd7260cc (pid=4530)
WARNING: failed to open status file: Resource temporarily unavailable
./partialscrub.sh: line 313: [: : integer expression expected
scrub cancelled
UUID:             48ed8a66-731d-499b-829e-dd07dd7260cc
Scrub resumed:    Sat Dec 19 14:36:28 2020
Status:           aborted
Duration:         0:26:55
Total to scrub:   87.02TiB
Rate:             153.60MiB/s
Error summary:    no errors found
Terminated
root@butter-server:/home/asif/bin/partialscrub#
```

Happy to provide any further information.

**Comment by dim-geo** [comment]
Can you paste your 

btrfs scrub status -dR /media/camino/
?

**Comment by yoasif** [comment]
Here you go:

```
root@butter-server:/home/asif/bin/partialscrub# btrfs scrub status -dR /media/camino/                                                                 
UUID:             48ed8a66-731d-499b-829e-dd07dd7260cc                                                                                                
scrub device /dev/sdh (id 4) history                                                                                                                  
Scrub resumed:    Sat Dec 19 16:44:54 2020                                                                                                            
Status:           aborted                                                                                                                             
Duration:         0:23:54                                                                                                                             
        data_extents_scrubbed: 3850259                                                                                                                
        tree_extents_scrubbed: 0                                                                                                                      
        data_bytes_scrubbed: 232622182400                                                                                                             
        tree_bytes_scrubbed: 0                                                                                                                        
        read_errors: 0
        csum_errors: 0
        verify_errors: 0
        no_csum: 1293
        csum_discards: 0
        super_errors: 0
        malloc_errors: 0
        uncorrectable_errors: 0
        unverified_errors: 0
        corrected_errors: 0
        last_physical: 234270294016
scrub device /dev/sdm (id 5)
        no stats available
scrub device /dev/sdd (id 7)
        no stats available
scrub device /dev/sdc (id 9)
        no stats available
scrub device /dev/sdj (id 10)
        no stats available
scrub device /dev/sdl (id 11)
        no stats available
scrub device /dev/sdf (id 12)
        no stats available
scrub device /dev/sda (id 14)
        no stats available
scrub device /dev/sdk (id 15)
        no stats available
scrub device /dev/sde (id 17)
        no stats available
scrub device /dev/sdn (id 18)
        no stats available
scrub device /dev/sdb (id 20)
        no stats available
scrub device /dev/sdo (id 21)
        no stats available
```

**Comment by dim-geo** [comment]
I don't see any obvious issue, maybe the scrub status failed temporarily due to many disks...
Did you try to rerun?
If it fails again, try to uncomment the echo current_position and increase the sleep counters to 30 if it falls.

**Comment by yoasif** [comment]
First tried to comment out the echo_position, which failed again after some time. 

Then modified the sleep 6 to sleep 30. Failed with a different error this time:

```
root@butter-server:/home/asif/bin/partialscrub# ./partialscrub.sh -d /media/camino/
filesystem:/media/camino/
raid56:on
percentage:100
frequency:2629744
scrub must be restarted:0
WARNING: RAID56 detected, not implemented
total data:42308876570624
data scrubbed:374147469312
status:aborted
limit:42683024039936
active device:0
scrub resumed on /dev/sdh, fsid 48ed8a66-731d-499b-829e-dd07dd7260cc (pid=6489)
Current Position:382911012864
Current Position:387702620160
WARNING: failed to open status file: Resource temporarily unavailable
./partialscrub.sh: line 313: [: !=: unary operator expected
scrub cancelled
UUID:             48ed8a66-731d-499b-829e-dd07dd7260cc
Scrub resumed:    Sat Dec 19 18:07:22 2020
Status:           aborted
Duration:         0:40:32
Total to scrub:   87.02TiB
Rate:             152.41MiB/s
Error summary:    no errors found
Terminated
```

**Comment by dim-geo** [comment]
Can you comment out set -x on the beginning?
It is weird because it seems to work a couple of times.

**Comment by yoasif** [comment]
Set -x was already commented out.

**Comment by dim-geo** [comment]
Sorry, I meant uncomment.

**Comment by yoasif** [comment]
I neglected to capture more data, but this is the last screenfull of output:

```
+ sleep 30
++ btrfs scrub status -R /media/camino/
++ grep data_bytes_scrubbed
++ awk '{print $2}'
+ current_position=2359250739200
+ echo 'Current Position:2359250739200'
+ '[' 2359250739200 -lt 42716223590400 ']'
++ btrfs scrub status /media/camino/
++ grep Status:
++ awk '{print $2}'
+ '[' running '!=' finished ']'
+ sleep 30
++ btrfs scrub status -R /media/camino/
++ grep data_bytes_scrubbed
++ awk '{print $2}'
+ current_position=2363143622656
+ echo 'Current Position:2363143622656'
+ '[' 2363143622656 -lt 42716223590400 ']'
++ btrfs scrub status /media/camino/
++ grep Status:
++ awk '{print $2}'
+ '[' running '!=' finished ']'
+ sleep 30
++ btrfs scrub status -R /media/camino/
++ grep data_bytes_scrubbed
++ awk '{print $2}'
WARNING: failed to open status file: Resource temporarily unavailable
+ current_position=
+ echo 'Current Position:'
+ '[' '' -lt 42716223590400 ']'
./partialscrub.sh: line 313: [: : integer expression expected
+ btrfs scrub cancel /media/camino/
+ btrfs scrub status /media/camino/
+ trap - SIGTERM
+ kill -- -2060
Terminated
```

**Comment by dim-geo** [comment]
Try this script:
https://github.com/dim-geo/partialscrub/blob/scrub_position/partialscrub.sh

If the script cannot acquire the current position, assume it's 0. At some point it will fetch the actual position and decide based on that.

**Comment by yoasif** [comment]
This is what ended up happening with the new script:

```
sudo ./partialscrub.sh -d /media/camino/
filesystem:/media/camino/
raid56:on
percentage:100
frequency:2629744
scrub must be restarted:0
WARNING: RAID56 detected, not implemented
total data:42144665149440
data scrubbed:3501522595840
status:aborted
limit:45646187745280
active device:0
scrub resumed on /dev/sdh, fsid 48ed8a66-731d-499b-829e-dd07dd7260cc (pid=11480)
WARNING: failed to open status file: Resource temporarily unavailable
WARNING: failed to open status file: Resource temporarily unavailable
./partialscrub.sh: line 321: [: !=: unary operator expected
scrub cancelled
UUID:             48ed8a66-731d-499b-829e-dd07dd7260cc
Scrub resumed:    Tue Dec 22 12:16:07 2020
Status:           aborted
Duration:         10:39:56
Total to scrub:   86.92TiB
Rate:             103.67MiB/s
Error summary:    read=9207073 csum=3
  Corrected:      103
  Uncorrectable:  9206973
  Unverified:     0
./partialscrub.sh: line 1: kill: (-11444) - No such process
```

**Comment by dim-geo** [comment]
You seem to have errors in your filesystem.

```
Error summary:    read=9207073 csum=3
  Corrected:      103
  Uncorrectable:  9206973
```
Please check your dmesg/syslog for errors. Is the device healthy? Maybe something fishy is happening in your filesystem. Maybe you need to replace some disk.

I updated the script in this branch to fetch the status as well in the same way.

**Comment by yoasif** [comment]
I know that there are errors in the fs, that is why I wanted to use this tool to do a scrub across disks sequentially. A full array scrub causes one of the disks to drop out (I know that I need to replace the disk), so I am trying to scrub the entire array sequentially as recommended for raid56 so I can more safely replace the failing disk.

**Comment by dim-geo** [comment]
I am not sure that you can find out which disk is faulty by scrub:

From https://lore.kernel.org/linux-btrfs/20200627032414.GX10769@hungrycats.org/#r
```
When there are csum failures, error counters of a random disk will be
incremented, not necessarily the disk that contains the corrupted blocks.
This makes it difficult or impossible to identify which disk in a raid5
array is corrupting data.
```

Can you check smart logs? Anyway the program is more robust now, I made another change in the branch above, you can check that as well.

**Comment by yoasif** [comment]
I know which disk is faulty, but if the disk drops out of the array, I can't correct any read errors if one of the mirrors containing the data has dropped out. I'm trying to scrub the array back into better shape.

I'll try the newest version of the script.

**Comment by dim-geo** [comment]
If you know which disk is faulty you should mount degraded and replace the disk. Scrubbing will not help now. Check the above link thoroughly please.

**Comment by yoasif** [comment]
This is a read heavy array so the data on the disks is likely fine, I just want to ensure that the data on the good disks is still good. I understand your reference, though.

I don't want to risk mounting degraded and losing data that is present on the flaky disk, if that clarifies. The array is not yet it in a "degraded" state.

**Comment by dim-geo** [comment]
You can replace a disk and minimize the reading from the faulty disk (-r option) The more you use a faulty disk, the chances increase that you will loose another disk and all data will be gone. Replace it and then scrub to see if there are losses. During replace all data will be read anyway, so scrubbing will be done because you use raid5. I hope that you use raid1 for metadata.

**Comment by yoasif** [comment]
The array is in a strange state because of a failed conversion to raid6 from raid10 (due to the dropped disk). I am using raid1c4 for metadata and I understand that I should have done a scrub after conversion to raid6 anyway, so I am scrubbing to get to a known state and an eventual conversion back to raid10 (somewhat scared off for now). Most of the errors above are read errors, not csum errors, and I have backups, so I'm not worried about this - I mostly want to see how resilient the array is during the conversion (maybe I can keep using raid6).

**Comment by dim-geo** [comment]
any news? Does it work better now?

**Comment by yoasif** [comment]
I have been using the new script. I also increased my device and eh_timeouts which has done wonders for my scrubs - no more read errors. I did this based on reading https://bugzilla.kernel.org/show_bug.cgi?id=93581.

I put partialscrub into my crontab and I don't think it has failed in a while. I don't know what caused the other issues, but at this point partialscrub works well enough for me - with the increased timeouts set on my disks.
